### PR TITLE
fix the logic of convert chunk segement into string

### PR DESCRIPTION
**Pull request by Yuyz0112**
#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message and code follows [Code of conduct](https://github.com/fastify/fastify/blob/master/CODE_OF_CONDUCT.md)

#### The bug

Currently we have this logic for the `asString` type parser's `onData` callback:
```js
body += chunk.toString()
```
This may cause the body string to be wrong because some character may have multi length buffer, for example chinese character:
```js
console.log(Buffer.from('啊')) // <Buffer e5 95 8a>
```
Since a multi length buffer may be cut into two chunks, `chunk.toString()` will make a broken string with symbol `&#65533`.

#### The fix

Just concat chunks in the callback and convert the whole buffer into string when transfer end.

#### The test

I'm not sure whether we can control the chunk size and then we could give a test to reproduce the bug. But if not, it seems hard to make a solid end to end test.

### CODE REVIEW

**Comment by mcollina** [comment]
I'd be good with a unit test.

I think a better way to handle this is to use `req.setEncoding('utf8')`, which will automatically convert in strings for us, without the need to keep all the buffers around.

**Comment by Yuyz0112** [comment]
ok, I would like to change the impl and give a unit test to it.
Where should I write this unit test? A new file or somewhere suggested?

**Comment by mcollina** [comment]
@Yuyz0112 add a new file to https://github.com/fastify/fastify/tree/master/test/internals. Just call the function, it would be easier.

@delvedor there should be no need to do `array.push`, as `.setEncoding('utf8')` should take care of this for us.

**Comment by Yuyz0112** [comment]
@mcollina
If I understood, you mean we should treat all asString type parser with encoding utf8 right?
I’m not familiar with other encodings and not sure wthether this may cause bugs in other encodings.

**Comment by mcollina** [comment]
@Yuyz0112 this is used by JSON parsing, and that is utf8 by default. Also the current `.toString()` assumes utf8.

**Comment by Yuyz0112** [comment]
@mcollina 
It looks like we also calcluate `receivedLength` in the `onData` callback:
```js
receivedLength += chunk.length
```
When using `.setEncoding('utf8')`, chunk will be string and we need to call `Buffer.from` to get the length of its buffer. Is this a faster approach than `array.push`?

**Comment by mcollina** [comment]
I'm fine in the length calculation to not be perfect. Those are a safety measure anyway.

**Comment by Yuyz0112** [comment]
request changed

File: test/helper.js [hunk 1]
```diff
@@ -268,6 +268,7 @@ module.exports.payloadMethod = function (method, t) {
       // Node errors for OPTIONS requests with a stream body and no Content-Length header
       if (upMethod !== 'OPTIONS') {
         var chunk = Buffer.allocUnsafe(1024 * 1024 + 1)
+        chunk.fill(0)
```
Comment (mcollina) [inline test/helper.js]: why this change?

Comment (Yuyz0112) [reply test/helper.js]: Because when we set `asString`, string length will be less or equal to the buffer length since some character consists of multi buffer.
This make `receivedLength <= limit` when we allocate random buffer.

In this test suite, I think the goal was to make sure every character consist of one buffer, so the `receivedLentgh` is controllable to tes the body limit feature.

Comment (mcollina) [reply test/helper.js]: Can you use `Buffer.alloc` instead?

Comment (Yuyz0112) [reply test/helper.js]: sure, done
